File: eprofiler-core/src/main/java/cz/encircled/eprofiler/DefaultMethodState.java
package cz.encircled.eprofiler;

import cz.encircled.eprofiler.core.ProfilerCore;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Kisel on 24.05.2016.
 */
public class DefaultMethodState implements MethodState {

    long id;

    long start;

    long totalTime;

    MethodState parent;

    List<MethodState> children = new ArrayList<>(4);

    private static String formatStackTrace(StackTraceElement traceElement) {
        return traceElement.getMethodName() + "#" + traceElement.getClassName();
    }

    @Override
    public long getId() {
        return id;
    }

    @Override
    public void setId(long id) {
        this.id = id;
    }

    @Override
    public MethodState getParent() {
        return parent;
    }

    @Override
    public void setParent(MethodState parent) {
        this.parent = parent;
    }

    @Override
    public void setStartTime(Long time) {
        start = time;
    }

    @Override
    public void end() {
        totalTime = Timer.now - start;
        if (totalTime < ProfilerCore.config().getMinDurationToLog() && parent != null) {
            // add merge

        } else {
            String parentId = parent == null ? "" : Long.toString(parent.getId());
            ProfilerCore.output().info(id + ":" + parentId + ":" + formatStackTrace(Thread.currentThread().getStackTrace()[2]) + ":" + totalTime);
        }

        if (hasParent()) {
            Profiler.state.set(parent);
        } else {
            Profiler.state.remove();
        }
    }

    @Override
    public void addParam(String value) {

    }

    @Override
    public boolean hasParent() {
        return parent != null;
    }

    @Override
    public MethodState addNested(MethodState state) {
        children.add(state);
        state.setParent(this);
        return state;
    }

}

File: eprofiler-core/src/main/java/cz/encircled/eprofiler/DefaultMethodState.kt
package cz.encircled.eprofiler

import cz.encircled.eprofiler.core.ProfilerCore
import java.util.*

/**
 * @author Kisel on 24.05.2016.
 */
class DefaultMethodState(override var id: Long) : MethodState {

    override var parent: MethodState? = null

    internal var start: Long = 0

    internal var totalTime: Long = 0

    internal var children: MutableList<MethodState> = ArrayList(4)

    private fun formatStackTrace(traceElement: StackTraceElement): String {
        return traceElement.methodName + "#" + traceElement.className
    }

    override fun setStartTime(time: Long?) {
        start = time!!
    }

    override fun end() {
        totalTime = Timer.now - start
        if (totalTime < ProfilerCore.configuration.minDurationToLog && parent != null) {
            // add merge

        } else {
            val parentId = if (parent == null) "" else java.lang.Long.toString(parent!!.id)
            ProfilerCore.outputWriter.info("$id : $parentId : ${formatStackTrace(Thread.currentThread().stackTrace[2])} : $totalTime")
        }

        if (hasParent()) {
            Profiler.state.set(parent)
        } else {
            Profiler.state.remove()
        }
    }

    override fun addParam(value: String) {

    }

    override fun hasParent(): Boolean {
        return parent != null
    }

    override fun addNested(state: MethodState): MethodState {
        children.add(state)
        state.parent = this
        return state
    }

}
